### PR TITLE
Adding support for configuring when matrix projects notify.

### LINKED ISSUE #198: 2.0.x Regression: one message emitted for each matrix conf + global job
Hi, before 2.0.x I received only one message telling the results of an executed job in a matrix job, and that's what I wanted in slack (when something fails, I check the details). Since 2.0.x version (I guess is connected to move the config to a post-build action), I'm getting spammed with one message per configuration in a matrix job plus one for the global job itself, which is quite annoying (for a simple job with built in 2 different environments, I get 3 messages). If including changes, then I get 6 messages in total, because I get one message with the changes for each configuration (which looks exactly the same) and one saying "No changes"  for the global job (I presume, the message just says "No Changes.").

Here is an example of before/after (ok, before having the changes listed was pointless as it always shown "No changes"):

before

now

### PR DESCRIPTION
**Pull request by kcboschert**
This adds support for users to configure whether a matrix project notifies:
1. Upon the completion of the parent job.
2. Upon the completion of each child job.
3. Both of the above.

This implementation solves #198 by pulling in similar logic from the [hipchat](https://github.com/jenkinsci/hipchat-plugin/commit/1310de16382dd266869201ea5a7cee724bffc061) and [email-ext](https://github.com/jenkinsci/email-ext-plugin/commit/906d076cb12443c5d1087e62ec428027e7a410de#diff-f739067328edce77ff51f8309b704d60) plugin.

While this does add the matrix-project as a dependency, this seems to be the pattern for how other publishers handle this, as seen by the plugins mentioned above. If this is acceptable, I'll add tests and request final review for merging.

Screen Shot 2019-04-02 at 2 17 07 PM

Fixes #198

### CODE REVIEW

**Comment by timja** [comment]
Hey, I'll take a look at this soon, could you add the needed tests please?

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 1]
```diff
@@ -470,6 +491,26 @@ public boolean prebuild(AbstractBuild<?, ?> build, BuildListener listener) {
         return super.prebuild(build, listener);
     }
 
+    public MatrixAggregator createAggregator(MatrixBuild matrixBuild, Launcher launcher, BuildListener buildListener) {
+        return new MatrixAggregator(matrixBuild, launcher, buildListener) {
+            @Override
+            public boolean startBuild() throws InterruptedException, IOException {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: IntelliJ says these exception declarations are unneeded?
```suggestion
            public boolean startBuild() {
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 2]
```diff
@@ -470,6 +491,26 @@ public boolean prebuild(AbstractBuild<?, ?> build, BuildListener listener) {
...
+                if (getMatrixTriggerMode().forParent) {
+                    return SlackNotifier.this.perform(this.build, this.launcher, this.listener);
+                }
+                return true;
+            }
+
+            @Override
+            public boolean endBuild() throws InterruptedException, IOException {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: IntelliJ says these exception declarations are unneeded?
```suggestion
            public boolean startBuild() {
```

**Comment by timja** [comment]
The build is currently failing for findbug violations:

```

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.FailedToParseSlackResponse(Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 40] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.MigratedCredentialDescription() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 108] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.NotificationFailed() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 220] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.NotificationFailedWithException(Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 176] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.SlackSendStepDisplayName() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 87] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.SlackSendStepValues(Object, Object, Object, Object, Object, Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 257] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.720Z] [INFO] The method name jenkins.plugins.slack.Messages.SlackSendStepValuesEmptyMessage() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 199] NM_METHOD_NAMING_CONVENTION

[2019-04-04T01:49:59.721Z] [INFO] 

[2019-04-04T01:49:59.721Z] 

[2019-04-04T01:49:59.721Z] 

[2019-04-04T01:49:59.721Z] To see bug detail using the Findbugs GUI, use the following command "mvn findbugs:gui"

[2019-04-04T01:49:59.721Z] 

[2019-04-04T01:49:59.721Z] 

[2019-04-04T01:49:59.721Z] 

[2019-04-04T01:49:59.721Z] [INFO] ------------------------------------------------------------------------

[2019-04-04T01:49:59.721Z] [INFO] BUILD FAILURE

[2019-04-04T01:49:59.721Z] [INFO] ------------------------------------------------------------------------

[2019-04-04T01:49:59.721Z] [INFO] Total time: 03:21 min

[2019-04-04T01:49:59.721Z] [INFO] Finished at: 2019-04-04T01:49:59Z

[2019-04-04T01:49:59.721Z] [INFO] ------------------------------------------------------------------------

[2019-04-04T01:49:59.721Z] [ERROR] Failed to execute goal org.codehaus.mojo:findbugs-maven-plugin:3.0.5:check (findbugs) on project slack: failed with 7 bugs and 0 errors -> [Help 1]

[2019-04-04T01:49:59.721Z] org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:findbugs-maven-plugin:3.0.5:check (findbugs) on project slack: failed with 7 bugs and 0 errors 
```

Honestly not sure why it's picking those up, all the other methods in there seem to start with a capital.

Any idea @casz?

**Comment by casz** [comment]
no idea why findbugs is looking at messages :laughing: I'd suggest updating parent pom and getting spotbugs instead.

Should be setup to ignore them.

File: findbugs-exclude.xml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<FindBugsFilter>
+	<Match>
+    <Class name="~.*\.Messages" />
+    <Bug pattern="NM_METHOD_NAMING_CONVENTION" />
```
Comment (casz) [inline findbugs-exclude.xml]: This shouldn't be needed...

**Comment by casz** [comment]
Please update parent POM to the least instead, should solve your issue

**Comment by kcboschert** [comment]
@casz Forgive my ignorance, but I'm not a Java developer. Can you give me more detail on what you're asking?

**Comment by casz** [comment]
Revert last addition and change the parent pom to the latest version: https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md

https://github.com/jenkinsci/slack-plugin/blob/a68a8b0c1044e7a909a091b4028569c874511d61/pom.xml#L6-L10
3.33 is already quite old please bump to latest

**Comment by kcboschert** [comment]
Updating to 3.42 gives the same errors with spotbugs:

```
[INFO] --- spotbugs-maven-plugin:3.1.11:check (default-cli) @ slack ---
[INFO] BugInstance size is 7
[INFO] Error size is 0
[INFO] Total bugs: 7
[ERROR] The method name jenkins.plugins.slack.Messages.FailedToParseSlackResponse(Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 40] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.MigratedCredentialDescription() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 108] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.NotificationFailed() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 220] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.NotificationFailedWithException(Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 176] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.SlackSendStepDisplayName() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 87] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.SlackSendStepValues(Object, Object, Object, Object, Object, Object) doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 257] NM_METHOD_NAMING_CONVENTION
[ERROR] The method name jenkins.plugins.slack.Messages.SlackSendStepValuesEmptyMessage() doesn't start with a lower case letter [jenkins.plugins.slack.Messages] At Messages.java:[line 199] NM_METHOD_NAMING_CONVENTION
[INFO]
```

**Comment by casz** [comment]
I think it is mocking that is pulling them into view

Could you try fixing them to lower case?

**Comment by casz** [comment]
@timja I think you would benefit from adding Travis... Jenkins agent is on the slow side...

Build #8 (Apr 8, 2019 9:20:05 PM) unacceptable waiting time.

**Comment by timja** [comment]
Tested regression, and light touch on the new feature (I don't know anything about matrix, but clicking through and running a build worked)

image

**Pull request merged**